glean-core/src/database/mod.rs

```diff
@@ -147,6 +166,60 @@ impl Database {
         }
     }
 
+    /// Checks if the 'user' lifetime metric is expired and, if so, removes it.
+    ///
+    /// ## Arguments
+    ///
+    /// * `storage_name`: The storage name to iterate over.
+    /// * `metric_id_slice`: The metric id to check.
+    ///
+    /// ## Return value
+    ///
+    /// Returns `true` if the metric is expired and should not be reported,
+    /// `false` otherwise.
+    ///
+    /// ## Panics
+    ///
+    /// This function will **not** panic on database errors.
+    fn check_and_remove_expired_user_lifetime(
+        &self,
+        storage_name: &str,
+        metric_id_slice: &[u8],
+    ) -> bool {
+        let metric_name = match str::from_utf8(metric_id_slice) {
+            Ok(metric_name) => metric_name.to_string(),
+            _ => return false,
+        };
+
+        match USER_LIFETIME_EXPIRATION_MAP.get(metric_name.as_str()) {
+            Some(&expiration) => {
+                let is_expired = *STARTUP_LOCAL_DATE > expiration;
+                if is_expired {
+                    // This should be fine to delete now, even though we may be iterating
+                    // through the database entries.
+                    if let Err(e) =
+                        self.remove_single_metric(Lifetime::User, storage_name, &metric_name)
```

Comment:
I think I know now why this is working fine: transactions!

Both the reading and iterating of the outside function as well as the deletion are running in their own transactions.
While the `remove_single_metric` immediately commits and thus removes data, the reader is in a separate transaction and thus will not see that changed data (= the fact that data is deleted).
That's why this is all safe and working

(rkv allows multiple open reader transactions, but only a single writer transaction)